### DISCUSSION TITLE
Rendering custom fields in export templates

### DISCUSSION THREAD

**Discussion by SandMouse** [netbox-community/netbox#6265] [Q&A] [answered]
### NetBox version

v2.11.1

### Python version

3.6

### Steps to Reproduce

1. Create export template "MyExport"

```
{% for address in queryset %}{% if address.cf_custom_0 %}

"{{address.cf_custom_1}}", "", "{{address.dns_name}}", "{{address.description}}", "{{address.cf_custom_2}}"

{% endif %}{% endfor %}
```

3. Create export: IPAM > IP Addresses > Export > "MyExport"

4. Save the file.

5. Open file, file is empty

### Expected Behavior

I expected that NetBox exports a file containg rows with the requested fields for records where the value of `cf_custom_0` is `True`.

### Observed Behavior

Exported file is empty.

### Note

If I remove the `if` `endif` it returns all ip addresses but custom fields remain empty. Built in fields are properly populated.

**Comment by jeremystretch** [comment, marked as answer]
`{% if address.cf_custom_0 %}` will never be true, because `cf_custom_0` isn't a property of the IPAddress model.

What specifically are you trying to do?

**Reply by SandMouse** [reply]
Jeremy, thanks for reaching out. When I lookup the defined fields in https://[fqdn]/ip-addresses/import/ I see a list of built-in fields as well as my custom defined fields. I used those field to create my export template. I assumed I could filter based on a custom field. I was unable to find this in the documentation. I'm trying to export IP addresses when a custom boolean field is `True`. But the custom fields stay empty. Even if I remove the `if` `endif`.

**Reply by candlerb** [reply]
For a custom field called "foo", `cf_foo` is only something you use in the REST API.

In the export templates, use `object.custom_field_data.foo` instead - e.g. `{% if address.custom_field_data.custom_0 %}`

EDIT: there is a [shortcut](https://github.com/netbox-community/netbox/blob/develop/netbox/netbox/models.py#L87-L92) `cf` as well, so `{% if address.cf.custom_0 %}` works as well.

**Reply by SandMouse** [reply]
@candlerb Thank you. I'll make adjustments. I've re-read the documentation and I see I've missed it. Thank you for pointing this out.

**Reply by SandMouse** [reply]
@candlerb Maybe you can hint me on sorting my output using a custom field. I'm trying:

```
{% for address in queryset|sort(attribute="custom_fields.custom_5,dns_name") %}{% if address.cf.custom_0 %}
...
{% endif %}{% endfor %}
```

I've tried `cf.custom_5`, `address.cf.custom_5` no joy. I tried `'ipam.models.ip.IPAddress.cf.custom_5` and `  ipaddress.cf.custom_5`

I can't find a working example. 

According to the Jinja2 documentation I should be able to do something like: 

```
{% for user in users|sort(attribute="name")
    |sort(reverse=true, attribute="age") %}
    ...
{% endfor %}
```

It seems I can't figure out how to properly address the attributes I want to use to sort my export using a custom field.

Can you point me in the right direction?

**Reply by candlerb** [reply]
The attribute is called  "custom_field_data" or "cf", not "custom_fields".

However, custom_field_data itself is a dict (of custom field name => custom field value).  Sorry, but I don't know how to get jinja2's [sort](https://jinja.palletsprojects.com/en/2.11.x/templates/#sort) to sort on a sub-value of an attribute.  Maybe it can't - there doesn't seem to be a general "sort by" filter.
